src/apps/rubicon/optimism/rubicon.bathToken.token-fetcher.ts

```diff
@@ -0,0 +1,124 @@
+
+    const appTokenDependencies = await this.appToolkit.getBaseTokenPrices(network);
+
+    const allTokenDependencies = [...appTokenDependencies];
+
+    // We will build a token object for each jar address, using data retrieved on-chain with Ethers
+    const tokens = await Promise.all(
+      bathTokens.map(async btAddress => {
+        // Instantiate a smart contract instance pointing to the jar token address
+        const contract = this.rubiconContractFactory.bathToken({
+          address: btAddress,
+          network,
+        });
+
+        // Request the symbol, decimals, ands supply for the jar token
+        const [symbol, decimals, supplyRaw] = await Promise.all([
+          multicall.wrap(contract).symbol(),
+          multicall.wrap(contract).decimals(),
+          multicall.wrap(contract).totalSupply(),
+        ]);
+
+        // Denormalize the supply
+        const supply = Number(supplyRaw) / 10 ** decimals;
+
+        // *******price ************
+        // A user can deposit base tokens like LOOKS or LQTY
+
+        // Request the underlying token address and ratio for the jar token
+        const [underlyingTokenAddressRaw, ratioRaw] = await Promise.all([
+          multicall
+            .wrap(contract)
+            .asset()
+            .catch(() => ''),
+          multicall
+            .wrap(contract)
+            .convertToAssets(BigNumber.from((1e18).toString()))
+            .catch(() => 'Convert to assets failed'),
+        ]);
+        // console.log('underlyingTokenAddressRaw', underlyingTokenAddressRaw, 'ratioRaw', ratioRaw.toString());
+
+        // Find the underlying token in our dependencies.
+        // Note: If it is not found, then we have not indexed the underlying token, and we cannot
+        // index the jar token since its price depends on the underlying token price.
+        const underlyingTokenAddress = underlyingTokenAddressRaw.toLowerCase();
+        const underlyingToken = allTokenDependencies.find(v => v.address === underlyingTokenAddress);
+        if (!underlyingToken) {
+          // console.log('No underlyingToken!', underlyingTokenAddress);
+
+          return null;
+        }
+        const tokens = [underlyingToken];
+
+        // Denormalize the price per share
+        const pricePerShare = Number(ratioRaw) / 10 ** 18;
+        const price = pricePerShare * underlyingToken.price;
+
+        // Create the token object
+        const token: AppTokenPosition = {
+          type: ContractType.APP_TOKEN,
+          appId,
+          groupId,
+          address: btAddress,
+          network,
+          symbol,
+          decimals,
+          supply,
+          tokens,
+          price,
+          pricePerShare,
+          // TODO: add any data props here
+          dataProps: {},
+          displayProps: {
+            label: '',
+            images: [],
+          },
+        };
+
+        return token;
+      }),
+    );
+
+    return tokens;
```

Comment:
Remember to filter out `null` values from the above `Promise.all` with `_.compact(tokens)`